Add/convert Doxygen comments for SideEffects and DoSimplifyExpressions visitors.

cole-barefoot opened the pull request:
@mbudiu-vmw: I added several TODO comments with questions that you may know the answer to.  If you think they are useful pointers to things we should address in the future, we can leave them in---if not, I'm happy to remove them.

mbudiu-vmw commented on frontends/p4/sideEffects.cpp:
```diff
@@ -22,14 +22,19 @@ namespace P4 {
 struct EvaluationOrder {
     ReferenceMap* refMap;
+
+    /// Output: The original expression with side-effecting sub-expressions
```
An expression whose evaluation will produce the same result as the original one, but where side-effects have been factored out such that each statement produces at most one side-effect.

mbudiu-vmw commented on frontends/p4/sideEffects.cpp:
```diff
@@ -61,13 +72,36 @@ struct EvaluationOrder {
     }
 };
 
+/** @brief Replace expressions containing constructor or method invocations
```
This also deals with short-circuit evaluation.

mbudiu-vmw commented on frontends/p4/sideEffects.cpp:
```diff
@@ -61,13 +72,36 @@ struct EvaluationOrder {
+ * left-value-ness and compile-time-contant-ness.
+ */
 class DismantleExpression : public Transform {
+    // TODO: It may be worth refactoring DismantleExpression to purely use
```
We actually want to precisely control the visit order, to reflect the order that the evaluation should be performed. I don't want to rely on the visitor order, which is not well-defined. This is the only way to really control the precise order: manual calls to visit and prune. See the compiler-design document too.

mbudiu-vmw commented on frontends/p4/sideEffects.cpp:
```diff
@@ -169,6 +203,7 @@ class DismantleExpression : public Transform {
         visit(expression->expr);
         auto left = result->final;
         CHECK_NULL(left);
+        // TODO: why clone here?
```
If you don't clone you have all sorts of weird problems (which are also a big pain to debug, BTW). For example, consider def-use analysis: the same expression has different uses and definitions depending of where it occurs in the program. So although something is syntactically identical, it should be represented by a different IR node. You will want different attributes computed for the two clones.

mbudiu-vmw commented on frontends/p4/sideEffects.cpp:
```diff
@@ -237,6 +273,7 @@ class DismantleExpression : public Transform {
             auto block = new IR::BlockStatement(*ifFalse);
             auto ifStatement = new IR::IfStatement(expression->srcInfo, cond, ifTrue, block);
             result->statements->push_back(ifStatement);
+            // TODO: result->addAssignment() returns a clone; why clone again?
```
If the path is used someplace else, one of the two users has to clone it.

mbudiu-vmw replied:
Also, better safe than sorry. Unless the contract for addAssignment is clear about this, I don't want to rely on it.

mbudiu-vmw commented on frontends/p4/sideEffects.cpp:
```diff
@@ -279,8 +317,11 @@ class DismantleExpression : public Transform {
+    // an expression.
+
+    // TODO: this could be made more precise, perhaps using LocationSets.
+    // TODO: surely there are other passes that compute reads/writes?  Perhaps
```
Unfortunately the pass that computes reads/writes depends on this one being run first. The reason is that you cannot compute reads if you don't know what side-effects occur first...

mbudiu-vmw commented on frontends/p4/sideEffects.cpp:
```diff
@@ -295,6 +336,8 @@ class DismantleExpression : public Transform {
         }
     };
 
+    // TODO: alias analysis would be generally useful... refactor this into
```
Alias analysis would use def-use information, which needs the information computed by this pass.

mbudiu-vmw commented on frontends/p4/sideEffects.h:
```diff
@@ -26,15 +26,23 @@ limitations under the License.
+ * invocations.
+ *
+ * The TypeMap and ReferenceMap arguments may be null, in which case every
+ * method call expression is counted.  With type information, invocations
```
"is assumed to have side-effects". Otherwise said, the analysis may be more conservative.

mbudiu-vmw commented on frontends/p4/sideEffects.h:
````diff
@@ -75,39 +87,41 @@ class SideEffects : public Inspector {
+ * - function is called with temporaries for all arguments
+ * - out and inout temporaries are copied to the saved left-values in order
+ *
+ * An assignment statement ```e = e1;``` is treated as if it is a call to a
````
I was too pedantic, this just means that the LHS is evaluated first.

For such a simple language, P4 has a really nasty semantics due to side-effects and out parameter. This pass was one of the hardest to write. After it has been run everything is much simpler.

cole-barefoot commented:
Thanks @mbudiu-vmw!  I revised based on your suggestions and removed the TODOs.  I also added a comment explaining why we implement a conservative alias analysis as part of this pass.
